core/src/apps/cardano/sign_tx.py

```diff
@@ -630,63 +644,64 @@ async def _show_stake_pool_registration_tx(
-) -> int:
-    total_amount = 0
-    for output in msg.outputs:
-        if output.address_parameters:
-            await _fail_or_warn_if_invalid_path(
-                ctx,
-                SCHEMA_ADDRESS,
-                output.address_parameters.address_n,
-                CHANGE_OUTPUT_PATH_NAME,
-            )
+async def _show_pool_owner(
+    ctx: wire.Context,
+    msg: CardanoSignTxInit,
+    keychain: seed.Keychain,
+    owner: CardanoPoolOwner,
+) -> None:
+    if owner.staking_key_path:
+        await _fail_or_warn_if_invalid_path(
+            ctx,
+            (SCHEMA_STAKING,),
+            owner.staking_key_path,
+            POOL_OWNER_STAKING_PATH_NAME,
+        )
 
-            await _show_change_output_staking_warnings(
-                ctx, keychain, output.address_parameters, output.amount
-            )
+    await confirm_stake_pool_owner(ctx, keychain, owner, msg.network_id)
 
-            if _should_hide_output(output.address_parameters.address_n, msg.inputs):
-                continue
 
-            address = derive_human_readable_address(
-                keychain, output.address_parameters, msg.protocol_magic, msg.network_id
-            )
-        else:
-            assert output.address is not None  # _validate_outputs
-            address = output.address
+def _validate_witness(
+    signing_mode: CardanoTxSigningMode,
+    witness: CardanoTxWitnessRequest,
+) -> None:
+    if signing_mode == CardanoTxSigningMode.POOL_REGISTRATION_AS_OWNER:
+        _ensure_no_payment_witness(witness)
 
-        total_amount += output.amount
 
-        if len(output.token_bundle) > 0:
-            await show_warning_tx_output_contains_tokens(ctx)
+def _ensure_no_payment_witness(witness: CardanoTxWitnessRequest) -> None:
+    """
+    We have a separate tx signing flow for stake pool registration because it's a
+    transaction where the witnessable entries (i.e. inputs, withdrawals, etc.)
+    in the transaction are not supposed to be controlled by the HW wallet, which
+    means the user is vulnerable to unknowingly supplying a witness for an UTXO
+    or other tx entry they think is external, resulting in the co-signers
+    gaining control over their funds (Something SLIP-0019 is dealing with for
+    BTC but no similar standard is currently available for Cardano). Hence we
+    completely forbid witnessing inputs and other entries of the transaction
+    except the stake pool certificate itself and we provide a witness only to the
+    user's staking key in the list of pool owners.
+    """
+    if not SCHEMA_STAKING_ANY_ACCOUNT.match(witness.path):
+        raise INVALID_STAKEPOOL_REGISTRATION_TX_WITNESSES
 
-        await confirm_sending(ctx, output.amount, output.token_bundle, address)
 
-    return total_amount
+async def _show_witness(
+    ctx: wire.Context,
+    witness_path: list[int],
+) -> None:
+    await _fail_or_warn_if_invalid_path(
+        ctx,
+        (SCHEMA_ADDRESS, SCHEMA_STAKING),
```

Comment:
I just realised that `SCHEMA_STAKING` is a subset of `SCHEMA_ADDRESS` so I can probably remove `SCHEMA_STAKING` from here and then revert `_fail_or_warn_if_invalid_path` to only take one schema parameter instead of a tuple.